js/ui/watermark.js

```diff
@@ -0,0 +1,237 @@
+// -*- mode: js; js-indent-level: 4; indent-tabs-mode: nil -*-
+
+const Clutter = imports.gi.Clutter;
+const Gio = imports.gi.Gio;
+const Lang = imports.lang;
+const St = imports.gi.St;
+const Background = imports.ui.background;
+const Main = imports.ui.main;
+const Monitor = imports.ui.monitor;
+const Tweener = imports.ui.tweener;
+
+const WATERMARK_SCHEMA = "org.gnome.shell.watermark";
+
+const Watermark = new Lang.Class({
+    Name: 'Watermark',
+
+    _init: function(bgManager) {
+        this._bgManager = bgManager;
+
+        this._watermarkFile = null;
+
+        this._settings = new Gio.Settings({ schema_id: WATERMARK_SCHEMA });
+
+        this._settings.connect('changed::watermark-file',
+                               Lang.bind(this, this._updateWatermark));
+        this._settings.connect('changed::watermark-size',
+                               Lang.bind(this, this._updateScale));
+        this._settings.connect('changed::watermark-position',
+                               Lang.bind(this, this._updatePosition));
+        this._settings.connect('changed::watermark-border',
+                               Lang.bind(this, this._updateBorder));
+        this._settings.connect('changed::watermark-always-visible',
+                               Lang.bind(this, this._updateVisibility));
+
+        this._textureCache = St.TextureCache.get_default();
+        this._textureCache.connect('texture-file-changed', Lang.bind(this,
+            function(cache, file) {
+                if (!this._watermarkFile || !this._watermarkFile.equal(file))
+                    return;
+
+                this._updateWatermarkTexture();
+            }));
+
+        this.actor = new St.Widget({ layout_manager: new Clutter.BinLayout(),
+                                     opacity: 0 });
+        bgManager._container.add_actor(this.actor);
+
+        this.actor.connect('destroy', Lang.bind(this, this._onDestroy));
+
+        let monitorIndex = bgManager._monitorIndex;
+        let constraint = new Monitor.MonitorConstraint({ index: monitorIndex,
+                                                         work_area: true});
+        this.actor.add_constraint(constraint);
+
+        this._bin = new St.Widget({ x_expand: true, y_expand: true });
+        this.actor.add_actor(this._bin);
+
+        this._settings.bind('watermark-opacity', this._bin, 'opacity',
+                            Gio.SettingsBindFlags.DEFAULT);
+        this._settings.connect('changed::watermark-opacity',
```

Comment:
Does not look like you need this (and the `_updateOpacity()` function altogether) because you use the binding.